### DISCUSSION TITLE
Allow S3 storage to receive addition `s3BucketEndpoint` parameter

### DISCUSSION THREAD

**Discussion by eltongonc** [directus/directus#16864] [Feature Requests]
I am trying to use Contabo Object Storage as a file storage for Directus. Their storage is compatible with S3 but uses a different endpoint URL structure.

when I use `STORAGE_S3_ENDPOINT=https://eu2.contabostorage.com/bucket-name`, I receive an error saying

```bash
Error code: UnknownEndpoint

Inaccessible host: `bucket-name.eu2.contabostorage.com' at port `undefined'.
 ```

This is the default behavior of S3 , which is to generate an host based on the params you give it.
But If we give it the `s3BucketEndpoint: true` as param, it will use the given `STORAGE_S3_ENDPOINT` instead.

To test this I've made a small node app with the `aws-sdk` with the following code:
```node
const S3 = require("aws-sdk/clients/s3");
require("dotenv").config();

const { STORAGE_URL, STORAGE_ACCESS_KEY, STORAGE_SECRET_KEY, STORAGE_BUCKET } =
  process.env;

// Connect to S3
const s3 = new S3({
  endpoint: STORAGE_URL,
  accessKeyId: STORAGE_ACCESS_KEY,
  secretAccessKey: STORAGE_SECRET_KEY,
  s3BucketEndpoint: true,
});

const keyName = "hello_world.txt";
// Create object upload promise
const uploadPromise = s3
  .putObject({
    Bucket: STORAGE_BUCKET,
    Key: keyName,
    Body: "Hello World!",
  })
  .promise();

uploadPromise.then(function (data) {
  console.log(
    "Successfully uploaded data to " + STORAGE_BUCKET + "/" + keyName
  );
});

```
Everything works perfectly fine, but if I comment out the `s3BucketEndpoint` will give me the same error received by directus.

### Proposal
It would be nice to have a check for the `s3BucketEndpoint`.

Perhaps in the `@directus/drive-s3` something like:
```node
this.$driver = new S3({
    accessKeyId: config.key,
    secretAccessKey: config.secret,
    s3BucketEndpoint: config.s3BucketEndpoint
    ...config,
});
```

**Comment by br41nslug** [comment]
arbitrary config options can be passed in the environment however getting the name right can be tricky.
`STORAGE_<LOCATION>_<CONFIG>` so something along the lines of `STORAGE_S3_S3_BUCKET_ENDPOINT="something"` could work :)

**Comment by eltongonc** [comment]
Why yes thank you @br41nslug! 

I've just tried adding `STORAGE_S3_s3BucketEndpoint: true` to my .env (in my case docker-compose.yml) and it worked. 

Reading through https://github.com/directus/directus/blob/bfe83d444e16a6cbc9a392092dc9a2e9f139b917/api/src/utils/get-config-from-env.ts gives me the impression that I can add any additional config as long as it has  `STORAGE_S3_` prefixed and `<CONFIG>`  has the exact casing as required for S3. 

Perhaps this is note-worthy for the docs?

**Comment by rijkvanzanten** [comment]
@br41nslug The updated s3 storage driver no longer automatically passes everything to S3, as it opened up quite a lot of unexpected side effects. We can make a new issue to officially support `BUCKET_ENDPOINT` as a setting (and then we can document it as well 🙂 ) once we figure out how this is configured in aws-sdk v3 🙂 

@eltongonc Any idea how this is set in `@aws-sdk/client-s3`?

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
